### ISSUE TITLE
Dataloader Crashes while using Multiple GPUs for ImageNet

### ISSUE THREAD

**Issue by AhmedHussKhalifa** [NVIDIA/DALI#4917]
### Describe the question.

Hey,
I am using the following dataloader with pytorch with 4 GPUs. In the following dataloader, I am generating for each sample in the training set different JPEG compressed verions. The code crashed after running 25 epochs. Is this beacuse of the device_memory_padding, host_memory_padding, preallocate_width_hint and preallocate_height_hint. Should I change them as my batch size is scaled by the numbers of the compressed values?

![image](https://github.com/NVIDIA/DALI/assets/9024871/f7e4b1b4-95bc-489c-bc80-9a7a9c457f34)

```
@pipeline_def
def create_dali_pipeline(data_dir, crop, size, shard_id, num_shards, NUM_QFs=0, dali_cpu=False, is_training=True):
    images, labels = fn.readers.file(file_root=data_dir,
                                     shard_id=shard_id,
                                     num_shards=num_shards,
                                     random_shuffle=is_training,
                                     pad_last_batch=True,
                                     name="Reader")
    dali_device = 'cpu' if dali_cpu else 'gpu'
    decoder_device = 'cpu' if dali_cpu else 'mixed'

    # ask nvJPEG to preallocate memory for the biggest sample in ImageNet for CPU and GPU to avoid reallocations in runtime
    device_memory_padding = 211025920 if decoder_device == 'mixed' else 0
    host_memory_padding = 140544512 if decoder_device == 'mixed' else 0
    # ask HW NVJPEG to allocate memory ahead for the biggest image in the data set to avoid reallocations in runtime
    preallocate_width_hint = 5980 if decoder_device == 'mixed' else 0
    preallocate_height_hint = 6430 if decoder_device == 'mixed' else 0
    if is_training: 
        images_org = fn.decoders.image_random_crop(images,
                                                device=decoder_device, output_type=types.RGB,
                                                device_memory_padding=device_memory_padding,
                                                host_memory_padding=host_memory_padding,
                                                preallocate_width_hint=preallocate_width_hint,
                                                preallocate_height_hint=preallocate_height_hint,
                                                random_aspect_ratio=[0.75, 4.0 / 3.0],
                                                random_area=[0.08, 1.0],
                                                num_attempts=100)
        resized_images = fn.resize(images_org,
                            device=dali_device,
                            resize_x=crop,
                            resize_y=crop,
                            interp_type=types.INTERP_TRIANGULAR)
        
        mirror = fn.random.coin_flip(probability=0.5)

        original = fn.crop_mirror_normalize(resized_images.gpu(),
                                        dtype=types.FLOAT,
                                        output_layout="CHW",
                                        crop=(crop, crop),
                                        mean=[0.485 * 255,0.456 * 255,0.406 * 255],
                                        std=[0.229 * 255,0.224 * 255,0.225 * 255],
                                        mirror=mirror)
        
        # List to store all the compressed samples
        compressed_samples = []
        compressed_samples.append(original)
        for qf in np.linspace(start=0, stop=100, num=NUM_QFs, endpoint=False):        
            compressed = fn.jpeg_compression_distortion(
                                                    resized_images.gpu(),
                                                    preserve=True,
                                                    quality=qf,
                                                    seed=0
                                                    )   
            compressed = fn.crop_mirror_normalize(compressed.gpu(),
                                            dtype=types.FLOAT,
                                            output_layout="CHW",
                                            crop=(crop, crop),
                                            mean=[0.485 * 255,0.456 * 255,0.406 * 255],
                                            std=[0.229 * 255,0.224 * 255,0.225 * 255],
                                            mirror=mirror)
            
            compressed_samples.append(compressed)            
        
        samples = fn.cat(*compressed_samples, axis=0)

    else:
        # This is for validation
        images = fn.decoders.image(images,
                                   device=decoder_device,
                                   output_type=types.RGB)
        images = fn.resize(
                            images_org,
                            device=dali_device,
                            size=size,
                            mode="not_smaller",
                            interp_type=types.INTERP_TRIANGULAR,
                            )
        mirror = False
        samples = fn.crop_mirror_normalize(images.gpu(),
                                        dtype=types.FLOAT,
                                        output_layout="CHW",
                                        crop=(crop, crop),
                                        mean=[0.485 * 255,0.456 * 255,0.406 * 255],
                                        std=[0.229 * 255,0.224 * 255,0.225 * 255],
                                        mirror=mirror)
    
    labels = labels.gpu()
    return samples, labels
```

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by JanuszL** [comment]
Hi @AhmedHussKhalifa,

Thank you for reaching out. The padding values are mostly for performance optimization (so the pipeline allocates enough memory ahead of time), you can remove them in your use case. Another thing you can do it to reduce the number of threads used by the pipeline and/or just reduce the batch size.

**Comment by AhmedHussKhalifa** [comment]
Thank you for your prompt response. I have a question regarding the need for padding values in my specific case, specifically for "device_memory_padding, host_memory_padding, preallocate_width_hint, and preallocate_height_hint." Could you please provide further clarification on their necessity?

Upon inspecting the image "train/n04418357n04418357_26036.JPEG" from ImageNet, which was the cause of the error mentioned earlier, I discovered that it has a high resolution. 

I have  removed the padding and re-run the experiment hopefully that this adjustment will prevent any further crashes, as reducing the batch size is not a viable option for me. I am adhering to the standard experimental setup, and decreasing the number of workers for the train dataloader would adversely affect the training speed. Could you please advise for another solution?

**Comment by JanuszL** [comment]
Hi @AhmedHussKhalifa,

The padding values tell the decoder to allocate at least XYZ memory when it tries to decode the image. In the usual conditions, the decoder allocates as much memory as it needs to decode a given image and holds it. When a bigger one is encountered it releases the old memory piece and allocates, newer and larger. To avoid this reallocation, if we know the data set well enough we can ask to allocate as much as it is needed to decode the biggest image and void slow reallocations. Image size hints work similarly but for the HW decoder.
As I mentioned earlier reducing the number of threads to 3-5 should not impact performance significantly. Feel free to try it out before ruling out this option.

**Comment by AhmedHussKhalifa** [comment]
Hey,

I have decreased the number of works twice and the code is table now. However, I believe the only way that I overcome any crushed is by resuming the training and saving each epoch as shown below.  

Could you please confirm that this is the right way to do and it will not affect my training process?



```
def save_ckp(state, is_best, checkpoint_dir, best_model_dir):
    f_path = checkpoint_dir / 'checkpoint.pt'
    torch.save(state, f_path)
    if is_best:
        best_fpath = best_model_dir / 'best_model.pt'
        shutil.copyfile(f_path, best_fpath)



def load_ckp(checkpoint_fpath, model, optimizer):
    checkpoint = torch.load(checkpoint_fpath)
    model.load_state_dict(checkpoint['state_dict'])
    optimizer.load_state_dict(checkpoint['optimizer'])
    return model, optimizer, checkpoint['epoch']


checkpoint = {
    'epoch': epoch + 1,
    'state_dict': model.state_dict(),
    'optimizer': optimizer.state_dict()
}
# For Each epoch
save_ckp(checkpoint, is_best, checkpoint_dir, model_dir)


....
# To resume training 
model = MyModel(*args, **kwargs)
optimizer = optim.Adam(model.parameters(), lr=learning_rate)
ckp_path = "path/to/checkpoint/checkpoint.pt"
model, optimizer, start_epoch = load_ckp(ckp_path, model, optimizer)

```

**Comment by AhmedHussKhalifa** [comment]
> ### Describe the question.
> Hey, I am using the following dataloader with pytorch with 4 GPUs. In the following dataloader, I am generating for each sample in the training set different JPEG compressed verions. The code crashed after running 25 epochs. Is this beacuse of the device_memory_padding, host_memory_padding, preallocate_width_hint and preallocate_height_hint. Should I change them as my batch size is scaled by the numbers of the compressed values?
> 
> ![image](https://user-images.githubusercontent.com/9024871/247134240-f7e4b1b4-95bc-489c-bc80-9a7a9c457f34.png)
> 
> ```
> @pipeline_def
> def create_dali_pipeline(data_dir, crop, size, shard_id, num_shards, NUM_QFs=0, dali_cpu=False, is_training=True):
>     images, labels = fn.readers.file(file_root=data_dir,
>                                      shard_id=shard_id,
>                                      num_shards=num_shards,
>                                      random_shuffle=is_training,
>                                      pad_last_batch=True,
>                                      name="Reader")
>     dali_device = 'cpu' if dali_cpu else 'gpu'
>     decoder_device = 'cpu' if dali_cpu else 'mixed'
> 
>     # ask nvJPEG to preallocate memory for the biggest sample in ImageNet for CPU and GPU to avoid reallocations in runtime
>     device_memory_padding = 211025920 if decoder_device == 'mixed' else 0
>     host_memory_padding = 140544512 if decoder_device == 'mixed' else 0
>     # ask HW NVJPEG to allocate memory ahead for the biggest image in the data set to avoid reallocations in runtime
>     preallocate_width_hint = 5980 if decoder_device == 'mixed' else 0
>     preallocate_height_hint = 6430 if decoder_device == 'mixed' else 0
>     if is_training: 
>         images_org = fn.decoders.image_random_crop(images,
>                                                 device=decoder_device, output_type=types.RGB,
>                                                 device_memory_padding=device_memory_padding,
>                                                 host_memory_padding=host_memory_padding,
>                                                 preallocate_width_hint=preallocate_width_hint,
>                                                 preallocate_height_hint=preallocate_height_hint,
>                                                 random_aspect_ratio=[0.75, 4.0 / 3.0],
>                                                 random_area=[0.08, 1.0],
>                                                 num_attempts=100)
>         resized_images = fn.resize(images_org,
>                             device=dali_device,
>                             resize_x=crop,
>                             resize_y=crop,
>                             interp_type=types.INTERP_TRIANGULAR)
>         
>         mirror = fn.random.coin_flip(probability=0.5)
> 
>         original = fn.crop_mirror_normalize(resized_images.gpu(),
>                                         dtype=types.FLOAT,
>                                         output_layout="CHW",
>                                         crop=(crop, crop),
>                                         mean=[0.485 * 255,0.456 * 255,0.406 * 255],
>                                         std=[0.229 * 255,0.224 * 255,0.225 * 255],
>                                         mirror=mirror)
>         
>         # List to store all the compressed samples
>         compressed_samples = []
>         compressed_samples.append(original)
>         for qf in np.linspace(start=0, stop=100, num=NUM_QFs, endpoint=False):        
>             compressed = fn.jpeg_compression_distortion(
>                                                     resized_images.gpu(),
>                                                     preserve=True,
>                                                     quality=qf,
>                                                     seed=0
>                                                     )   
>             compressed = fn.crop_mirror_normalize(compressed.gpu(),
>                                             dtype=types.FLOAT,
>                                             output_layout="CHW",
>                                             crop=(crop, crop),
>                                             mean=[0.485 * 255,0.456 * 255,0.406 * 255],
>                                             std=[0.229 * 255,0.224 * 255,0.225 * 255],
>                                             mirror=mirror)
>             
>             compressed_samples.append(compressed)            
>         
>         samples = fn.cat(*compressed_samples, axis=0)
> 
>     else:
>         # This is for validation
>         images = fn.decoders.image(images,
>                                    device=decoder_device,
>                                    output_type=types.RGB)
>         images = fn.resize(
>                             images_org,
>                             device=dali_device,
>                             size=size,
>                             mode="not_smaller",
>                             interp_type=types.INTERP_TRIANGULAR,
>                             )
>         mirror = False
>         samples = fn.crop_mirror_normalize(images.gpu(),
>                                         dtype=types.FLOAT,
>                                         output_layout="CHW",
>                                         crop=(crop, crop),
>                                         mean=[0.485 * 255,0.456 * 255,0.406 * 255],
>                                         std=[0.229 * 255,0.224 * 255,0.225 * 255],
>                                         mirror=mirror)
>     
>     labels = labels.gpu()
>     return samples, labels
> ```
> 
> ### Check for duplicates
> * [x]  I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by JanuszL** [comment]
Hi @AhmedHussKhalifa,

I cannot opine the model saving, in the case of DALI itself, it doesn't provide an explicit checkpointing ability. However, what you can do is rebuild the DALI pipeline for each epoch with a different (predefined) seed and save it. When you want to resume the training just create the pipeline with the saved seed.

**Comment by AhmedHussKhalifa** [comment]
Hey @JanuszL,

Just to confirm the setup that you have suggested to presume training after a sudden stop, In each epoch, I do the following: The first thing is to always save the latest model weights and optimizer parameters (in my case, I use SGD); secondly, I rebuild the DALI pipeline with a different seed that is predefined. Could you please clarify why I need to do this?

Another thing I have observed is that the GPU memory is slightly increasing after each epoch. Could you tell me how to solve this problem in the case shown above and why it is happening?

**Comment by JanuszL** [comment]
Hi @AhmedHussKhalifa,

You can do something like this in the epoch loop:
```
for epoch_number in range (number_of_epochs):
    train_pipe = create_dali_pipeline(batch_size=args.batch_size,
                                      num_threads=args.workers,
                                      device_id=args.local_rank,
                                      seed=12 + args.local_rank + epoch_number,
                                      data_dir=traindir,
                                      crop=crop_size,
                                      size=val_size,
                                      dali_cpu=args.dali_cpu,
                                      shard_id=args.local_rank,
                                      num_shards=args.world_size,
                                      is_training=True)
    train_pipe.build()
    train_loader = DALIClassificationIterator(train_pipe, reader_name="Reader",
                                              last_batch_policy=LastBatchPolicy.PARTIAL,
                                              auto_reset=True)
```

> Another thing I have observed is that the GPU memory is slightly increasing after each epoch.

With the recent DALI version which uses a memory pool, the GPU memory should hit the water line pretty fast. The only thing that case allocates additional memory over time is the image decoder (it uses a different pool). Still, consumption should asymptotically converge to one value.

**Comment by AhmedHussKhalifa** [comment]
Thank you for your help, @JanuszL .

**Comment by AhmedHussKhalifa** [comment]
Heu @JanuszL,

Sorry for reopening this issue, but I got a new bug while running the training (shown below). I have followed all that you have suggested by decreasing the number of workers, but it seems the training is not stable.

![image](https://github.com/NVIDIA/DALI/assets/9024871/c9605a8d-71ba-465e-868a-21b2961377ca)

**Comment by JanuszL** [comment]
Hi @AhmedHussKhalifa,

It seems the operator tries to acquire a lot of memory (900MB). What is the batch size you use? What is the data set? I think in most cases it should not use such an amount of memory unless the batch size is really big/the images you have in the data set have high resolution.

**Comment by JanuszL** [comment]
Closing as stalled.
